tedana/decay.py

```diff
@@ -460,3 +462,71 @@ def fit_decay_ts(data, tes, mask, adaptive_mask, fittype):
         report = False
 
     return t2s_limited_ts, s0_limited_ts, t2s_full_ts, s0_full_ts
+
+
+def rmse_of_fit_decay_ts(
+    *,
+    data: np.ndarray,
+    tes: List[float],
+    adaptive_mask: np.ndarray,
+    t2s: np.ndarray,
+    s0: np.ndarray,
+    fitmode: Literal["all", "ts"],
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Estimate model fit of voxel- and timepoint-wise monoexponential decay models to ``data``.
+
+    Parameters
+    ----------
+    data : (S x E x T) :obj:`numpy.ndarray`
+        Multi-echo data array, where `S` is samples, `E` is echos, and `T` is time.
+    tes : (E,) :obj:`list`
+        Echo times.
+    adaptive_mask : (S,) :obj:`numpy.ndarray`
+        Array where each value indicates the number of echoes with good signal for that voxel.
+        This mask may be thresholded; for example, with values less than 3 set to 0.
+        For more information on thresholding, see :func:`~tedana.utils.make_adaptive_mask`.
+    t2s : (S [x T]) :obj:`numpy.ndarray`
+        Voxel-wise (and possibly volume-wise) T2* estimates from
+        :func:`~tedana.decay.fit_decay_ts`.
+    s0 : (S [x T]) :obj:`numpy.ndarray`
+        Voxel-wise (and possibly volume-wise) S0 estimates from :func:`~tedana.decay.fit_decay_ts`.
+    fitmode : {"fit", "all"}
+        Whether the T2* and S0 estimates are volume-wise ("fit") or not ("all").
+
+    Returns
+    -------
+    rmse_map : (S,) :obj:`numpy.ndarray`
+        Mean root mean squared error of the model fit across all volumes at each voxel.
+    rmse_timeseries : (T,) :obj:`numpy.ndarray`
+        Mean root mean squared error of the model fit across all voxels at each timepoint.
+    rmse_sd_timeseries : (T,) :obj:`numpy.ndarray`
+        Standard deviation of root mean squared error of the model fit across all voxels at each
+        timepoint.
+    """
+    n_samples, _, n_vols = data.shape
+    tes = np.array(tes)
+
+    rmse = np.zeros([n_samples, n_vols])
+    for i_voxel in tqdm(range(n_samples)):
+        n_good_echoes = adaptive_mask[i_voxel]
+
+        data_voxel = data[i_voxel, :n_good_echoes, :]
+
+        if fitmode == "all":
+            s0_voxel = np.full(data_voxel.shape[-1], s0[i_voxel])
+            t2s_voxel = np.full(data_voxel.shape[-1], t2s[i_voxel])
+        else:
+            s0_voxel = s0[i_voxel, :]
+            t2s_voxel = t2s[i_voxel, :]
+
+        predicted_data = monoexponential(
+            tes=tes[:n_good_echoes, None],
+            s0=s0_voxel[None, :],
+            t2star=t2s_voxel[None, :],
+        )
+        rmse[i_voxel, :] = np.sqrt(np.mean((data_voxel - predicted_data) ** 2, axis=0))
```

Comment:
This should be a more efficient solution that removes the need to use `tqdm`
I tried this on one dataset and `rmse` is identical for voxels with 2 or more good echoes. I decided to give no values (`nan`) for only 1 good echo since there shouldn't be any s0 and t2s estimate for those voxels and a fit to no model isn't informative. Note that this also required adding `import numpy.matlib` (Not sure why just numpy wasn't enough)

```suggestion
    rmse = np.full([n_samples, n_vols], np.nan, dtype=np.float32)
    # n_good_echoes interates from 2 through the number of echoes
    #   0 and 1 are excluded because there aren't T2* and S0 estimates
    #   for less than 2 voxels. 2 voxels will have a bad estimate so consider
    #   how/if we want to distinguish those
    for n_good_echoes in range(2, len(tes) + 1):
        # a boolean mask for voxels with a specific num of good echoes
        use_vox = adaptive_mask == n_good_echoes
        data_echo = data[use_vox, :n_good_echoes, :]
        if fitmode == "all":
            s0_echo = np.matlib.repmat(s0[use_vox].T, n_vols, 1).T
            t2s_echo = np.matlib.repmat(t2s[use_vox], n_vols, 1).T
        elif fitmode == "fit":
            s0_echo = s0[use_vox, :]
            t2s_echo = t2s[use_vox, :]
        # TODO add an else if fitmode is neither.
        # TODO Might also want to check if s0 and t2 are the right dimensions for fitmode option

        predicted_data = np.full([use_vox.sum(), n_good_echoes, n_vols], np.nan, dtype=np.float32)
        # Need to loop by echo since monoexponential can take either single vals for s0 and t2star
        #   or a single TE value. We could expand that func, but this is a functional solution
        for echo_num in range(n_good_echoes):
            predicted_data[:, echo_num, :] = monoexponential(
                tes=tes[echo_num],
                s0=s0_echo,
                t2star=t2s_echo,
            )
        rmse[use_vox, :] = np.sqrt(np.mean((data_echo - predicted_data) ** 2, axis=1))
```